cli/flox-rust-sdk/src/providers/buildenv.rs

```diff
@@ -94,77 +107,347 @@ pub trait BuildEnv {
+    ///    <https://github.com/flox/nixpkgs>, which enables building packages
+    ///    without common evaluation checks, such as unfree and broken.
+    /// 2. constructing the attribute path to build the package,
+    ///    i.e. `legacyPackages.<locked system>.<attr_path>`,
+    ///    as [LockedPackageCatalog::attr_path] is incomplete.
+    /// 3. building the package with essentially
+    ///    `nix build <flox-nixpkgs-url>#<resolved attr path>^*`,
+    ///    which will realise the locked output paths.
+    ///    We set `--option pure-eval true` to improve reproducibility
+    ///    of the locked outputs, and allow the use of the eval-cache
+    ///    to avoid costly re-evaluations.
+    ///
+    /// IMPORTANT/TODO: As custom catalogs, with non-nixpkgs packages are in development,
+    /// this function is currently assumes that the package is from the nixpkgs base-catalog.
+    /// Currently the type is distinguished by the [LockedPackageCatalog::locked_url].
+    /// If this does not indicate a nixpkgs package, the function will currently panic!
+    fn realise_nixpkgs(&self, locked: &LockedPackageCatalog) -> Result<(), BuildEnvError> {
+        // check if all store paths are valid, if so, return without eval
+        let all_valid = self.check_store_path(locked.outputs.values())?;
+
+        if all_valid {
+            return Ok(());
         }
 
-        // build the environment
         let mut nix_build_command = self.base_command();
 
+        // for now assume the plugin is located relative to the pkgdb binary
+        // <pkgdb>
+        // ├── bin
+        // │   └── pkgdb
+        // └── lib
+        //     └── wrapped-nixpkgs-input.(so|dylib)
+        {
+            let pkgdb_lib_dir = Path::new(&*PKGDB_BIN)
+                .ancestors()
+                .nth(2)
+                .expect("pkgdb is in '<store-path>/bin'")
+                .join("lib")
+                .join(format!("wrapped-nixpkgs-input{}", env!("libExt")));
+            nix_build_command.args([
+                "--option",
+                "extra-plugin-files",
+                &pkgdb_lib_dir.to_string_lossy(),
+            ]);
+        }
+
+        let installable = {
+            let mut locked_url = locked.locked_url.to_string();
+            if let Some(revision_suffix) =
+                locked_url.strip_prefix("https://github.com/flox/nixpkgs?rev=")
+            {
+                locked_url = format!("flox-nixpkgs:v0/flox/{revision_suffix}");
+            } else {
+                todo!(
+                    "Building non-nixpkgs catalog packages is not yet supported.\n\
+                    Pending implementation and decisions regarding representation in the lockfile"
+                );
+            }
+
+            // build all out paths
+            let attrpath = format!("legacyPackages.{}.{}^*", locked.system, locked.attr_path);
+
+            format!("{}#{}", locked_url, attrpath)
+        };
+
+        nix_build_command.arg("build");
```

Comment:
followup suggestion: for free packages, can we realise them via substitution without getting the plugin and a `flox-nixpkgs` flakeref involved at all?

Reply:
yes, and as of now we actually do, using `nix build <store paths...>` to check for the existence of paths locally, or via substitution